

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Input data and preprocessing &mdash; adabmDCA 1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=f2a433a1"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Implementation" href="implementation.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            adabmDCA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Input data and preprocessing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-data">Input data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preprocessing">Preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#preprocessing-pipeline">Preprocessing pipeline</a></li>
<li class="toctree-l3"><a class="reference internal" href="#computing-the-importance-weights">Computing the importance weights</a></li>
<li class="toctree-l3"><a class="reference internal" href="#pseudo-count-and-reweighted-statistics">Pseudo count and reweighted statistics</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="implementation.html">Implementation</a></li>
<li class="toctree-l1"><a class="reference internal" href="applications.html">Applications</a></li>
<li class="toctree-l1"><a class="reference internal" href="boltzmann_learning.html">Boltzmann learning of biological models</a></li>
<li class="toctree-l1"><a class="reference internal" href="quicklist.html">Quicklist</a></li>
<li class="toctree-l1"><a class="reference internal" href="script_arguments.html">Script Arguments</a></li>
<li class="toctree-l1"><a class="reference internal" href="adabmDCApy.html">adabmDCApy APIs</a></li>
<li class="toctree-l1"><a class="reference internal" href="bibliography.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">adabmDCA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Input data and preprocessing</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/preprocessing.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="input-data-and-preprocessing">
<h1>Input data and preprocessing<a class="headerlink" href="#input-data-and-preprocessing" title="Link to this heading"></a></h1>
<section id="input-data">
<h2>Input data<a class="headerlink" href="#input-data" title="Link to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">adabmDCA</span> <span class="pre">2.0</span></code> takes as input a multiple sequence alignment (<a class="reference external" href="https://en.wikipedia.org/wiki/Multiple_sequence_alignment">MSA</a>) of aligned amino acid or nucleotide sequences, usually forming a protein or RNA family. DCA implementations require the data to be saved in FASTA format <span id="id1">[<a class="reference internal" href="bibliography.html#id12" title="W R Pearson and D J Lipman. Improved tools for biological sequence comparison. Proceedings of the National Academy of Sciences of the United States of America, 85(8):2444–2448, 1988. URL: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC280013/.">Pearson and Lipman, 1988</a>]</span>.</p>
<p><code class="docutils literal notranslate"><span class="pre">adabmDCA</span> <span class="pre">2.0</span></code> implements the three default alphabets shown in table <a class="reference internal" href="#alphabets">Alphabets</a>, but the user can specify an ad-hoc alphabet as far as it is compatible with the input MSA.</p>
<table class="docutils align-center" id="alphabets">
<thead>
<tr class="row-odd"><th class="head text-center"><p></p></th>
<th class="head text-center"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-center"><p><strong>protein</strong></p></td>
<td class="text-center"><p><code class="docutils literal notranslate"><span class="pre">-,</span> <span class="pre">A,</span> <span class="pre">C,</span> <span class="pre">D,</span> <span class="pre">E,</span> <span class="pre">F,</span> <span class="pre">G,</span> <span class="pre">H,</span> <span class="pre">I,</span> <span class="pre">K,</span> <span class="pre">L,</span> <span class="pre">M,</span> <span class="pre">N,</span> <span class="pre">P,</span> <span class="pre">Q,</span> <span class="pre">R,</span> <span class="pre">S,</span> <span class="pre">T,</span> <span class="pre">V,</span> <span class="pre">W,</span> <span class="pre">Y</span></code></p></td>
</tr>
<tr class="row-odd"><td class="text-center"><p><strong>rna</strong></p></td>
<td class="text-center"><p><code class="docutils literal notranslate"><span class="pre">-,</span> <span class="pre">A,</span> <span class="pre">C,</span> <span class="pre">G,</span> <span class="pre">U</span></code></p></td>
</tr>
<tr class="row-even"><td class="text-center"><p><strong>dna</strong></p></td>
<td class="text-center"><p><code class="docutils literal notranslate"><span class="pre">-,</span> <span class="pre">A,</span> <span class="pre">C,</span> <span class="pre">G,</span> <span class="pre">T</span></code></p></td>
</tr>
</tbody>
</table>
<p>An example of a FASTA file format is shown in Figure <a class="reference internal" href="#fig-example-fasta"><span class="std std-ref">Example FASTA</span></a>. In particular, <code class="docutils literal notranslate"><span class="pre">adabmDCA</span> <span class="pre">2.0</span></code> correctly handles FASTA files in which line breaks within a sequence are present.</p>
<figure class="align-center" id="fig-example-fasta">
<a class="reference internal image-reference" href="_images/example_fasta_2.png"><img alt="Fasta example" src="_images/example_fasta_2.png" style="width: 715.0px; height: 258.0px;" />
</a>
<figcaption>
<p><span class="caption-text">Example of protein sequences formatted in FASTA format.</span><a class="headerlink" href="#fig-example-fasta" title="Link to this image"></a></p>
</figcaption>
</figure>
</section>
<section id="preprocessing">
<h2>Preprocessing<a class="headerlink" href="#preprocessing" title="Link to this heading"></a></h2>
<section id="preprocessing-pipeline">
<h3>Preprocessing pipeline<a class="headerlink" href="#preprocessing-pipeline" title="Link to this heading"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">adabmDCA</span> <span class="pre">2.0</span></code> code applies the following preprocessing pipeline to
the input MSA:</p>
<ul class="simple">
<li><p>Remove the sequences having some tokens not included in the default
alphabet;</p></li>
<li><p>Compute the importance weights for the sequences in the MSA;</p></li>
<li><p>Apply a pseudocount to compute the MSA stastistics.</p></li>
</ul>
<p>Their precise implementation is described in the following.</p>
</section>
<section id="computing-the-importance-weights">
<h3>Computing the importance weights<a class="headerlink" href="#computing-the-importance-weights" title="Link to this heading"></a></h3>
<p>The sequence weights are computed to mitigate as much as possible the systematic biases in the data, such as correlations due to the phylogeny or over-representation of some regions of the sequence space because of a sequencing bias.</p>
<p>Given an MSA of <span class="math notranslate nohighlight">\(M\)</span> sequences, to compute the importance weight of each sequence <span class="math notranslate nohighlight">\(\pmb a^{(m)}\)</span>, <span class="math notranslate nohighlight">\(m=1, \dots, M\)</span>, we consider <span class="math notranslate nohighlight">\(N^{(m)}\)</span> as the number of sequences in the dataset having Hamming distance from <span class="math notranslate nohighlight">\(\pmb a^{(m)}\)</span> smaller or equal to <span class="math notranslate nohighlight">\(0.2 \cdot L\)</span> (this threshold can be tuned by the user). Then, the importance  weight of <span class="math notranslate nohighlight">\(\pmb{a}^{(m)}\)</span> will be</p>
<div class="math notranslate nohighlight" id="equation-eqn-weights">
<span class="eqno">(1)<a class="headerlink" href="#equation-eqn-weights" title="Link to this equation"></a></span>\[
w^{(m)} = \frac{1}{N^{(m)}}.
\]</div>
<p>This reweighting allows us to give less importance to sequences found in very densely populated regions of the sequence space while enhancing the importance of isolated sequences.</p>
</section>
<section id="pseudo-count-and-reweighted-statistics">
<h3>Pseudo count and reweighted statistics<a class="headerlink" href="#pseudo-count-and-reweighted-statistics" title="Link to this heading"></a></h3>
<p>DCA models are trained to reproduce the one and two-site frequencies of the empirical data. To compute these, we introduce in the computation of the empirical statistics a small parameter <span class="math notranslate nohighlight">\(\alpha\)</span>, called pseudo count, that allows us to deal with unobserved (pairs of) symbols in one (or two) column(s) of the MSA. The one and two-site frequencies are given by</p>
<div class="math notranslate nohighlight" id="equation-eqn-freqs-pseudocount">
<span class="eqno">(2)<a class="headerlink" href="#equation-eqn-freqs-pseudocount" title="Link to this equation"></a></span>\[\begin{split}
    f_i(a) &amp;= (1 - \alpha) f^{\mathrm{data}}_i(a) + \frac{\alpha}{q}, \\
    f_{ij}(a, b) &amp;= (1 - \alpha) f^{\mathrm{data}}_{ij}(a, b) + \frac{\alpha}{q^2}\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(f_i^{\mathrm{data}}(a)\)</span> and <span class="math notranslate nohighlight">\(f_{ij}^{\mathrm{data}}(a, b)\)</span> are computed from the MSA as in Eq. <a class="reference internal" href="boltzmann_learning.html#equation-eqn-freqs">(3)</a>.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="implementation.html" class="btn btn-neutral float-right" title="Implementation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Lorenzo Rosset.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>